cmds/dump.c

```diff
@@ -17,61 +17,214 @@
 
 #include <hal/hal.h>
 #include <lib/lib.h>
+#include <syspage.h>
+
+#include <phfs/phfs.h>
 
 
 static void cmd_dumpInfo(void)
 {
-	lib_printf("dumps memory, usage: dump <addr>");
+	lib_printf("dumps memory, usage: dump [-F|-r <phfs>] <addr> [<size>]");
+}
+
+
+static unsigned int region_validator(addr_t start, addr_t end, addr_t curr)
+{
+	unsigned int attr = 0;
+
+	(void)start;
+	(void)end;
+
+	if (syspage_mapRangeCheck(curr, curr, &attr)) {
+		attr = (attr & mAttrRead);
+	}
+
+	return attr != 0;
+}
+
+
+static void region_hexdump(addr_t start, addr_t end, addr_t offp, u8 align, unsigned int (*validator)(addr_t, addr_t, addr_t))
+{
+	unsigned int col;
+	static const int cols = 16;
+	addr_t ptr, offs = start;
+
+	if (align != 0) {
+		offp = (offp / cols) * cols;
+		offs = (offs / cols) * cols;
+	}
+
+	while (offs < end) {
+		lib_printf("0x%08x | ", (u32)offp);
+
+		/* Print byte values */
+		for (col = 0; col < cols; ++col) {
+			ptr = offs + col;
+
+			if ((ptr < start) || (ptr >= end)) {
+				lib_printf("   ");
+				continue;
+			}
+
+			if ((validator != NULL) && (validator(start, end, ptr) == 0)) {
+				lib_printf("XX ");
+			}
+			else {
+				lib_printf("%02x ", *(u8 *)ptr);
+			}
+		}
+		lib_printf("| ");
+
+		/* Print "printable" representation */
+		for (col = 0; col < cols; ++col) {
+			ptr = offs + col;
+
+			if ((ptr < start) || (ptr >= end)) {
+				lib_printf(" ");
+				continue;
+			}
+
+			if ((validator != NULL) && (validator(start, end, ptr) == 0)) {
+				lib_printf("X");
+			}
+			else if (lib_isprint(*(u8 *)ptr)) {
+				lib_printf("%c", *(u8 *)ptr);
+			}
+			else {
+				lib_printf(".");
+			}
+		}
+
+		lib_printf("\n");
+		offp += cols;
+		offs += cols;
+	}
+}
+
+
+static void print_hline(void)
+{
+	unsigned int col;
+	for (col = 0; col < 79; ++col) {
+		lib_consolePutc('-');
+	}
+	lib_printf("\n");
```

Comment:
Just a suggestion, somewhat faster approach, we don't have to worry about stack in plo that much
```suggestion
	char buff[80];
	hal_memset(buff, '-', sizeof(buff) - 1);
	buff[sizeof(buff) - 1] = '\n';
	lib_printf(buff);
```

Reply:
`printf()` [underneath calls](https://github.com/phoenix-rtos/plo/blob/master/lib/printf.c#L25-L29) `lib_consolePutc()` so it's not the fastest and is stack hungry, but well ... I missed that last `lib_printf("\n")` can be replaced with `lib_consolePutc('\n')` :-) Thanks in advance for pointing it out. I'll change it like so, and other calls to `printf()` that does not use formatting I'll change to `lib_consolePuts()`, this is what compiler do for `printf`s if it detects no formating it replaces with puts/putc respectively.